packaging/test_plugin_unified_storage_tiering.py

```diff
@@ -760,6 +767,86 @@ def test_put_and_get(self):
 
+    def test_open_read_close_triggers_restage__issue_303(self):
+        with storage_tiering_configured():
+            IrodsController().restart(test_mode=True)
+            filename = "test_open_read_close_triggers_restage__issue_303"
+            try:
+                logical_path = "/".join([self.user1.session_collection, filename])
+                lib.create_local_testfile(filename)
+                self.user1.assert_icommand(["iput", "-R", "ufs0", filename, logical_path])
+
+                # Tier out the object...
+                self.user1.assert_icommand(["ils", "-L", logical_path], "STDOUT", self.tier0)
+                time.sleep(5)
+                invoke_storage_tiering_rule()
+                delay_assert_icommand(self.user1, f"ils -L {logical_path}", 'STDOUT', self.tier1)
+
+                # And then again...
+                time.sleep(15)
+                invoke_storage_tiering_rule()
+                delay_assert_icommand(self.user1, f"ils -L {logical_path}", 'STDOUT', self.tier2)
+
+                # Now see if the object is restaged appropriately on open/read/close.
+                self.user1.assert_icommand(["irods_test_read_object", logical_path], "STDOUT")
+                # Wait for the object to be trimmed before checking destination to avoid timing issues.
+                lib.delayAssert(
+                    lambda: lib.replica_exists_on_resource(self.user1, logical_path, self.tier2) == False)
+                self.assertTrue(lib.replica_exists_on_resource(self.user1, logical_path, self.tier1))
+
+            finally:
+                if os.path.exists(filename):
+                    os.unlink(filename)
+                self.user1.assert_icommand(["irm", "-f", logical_path])
+
+    def creating_data_object_does_not_trigger_restage_test_impl(self, create_command, logical_path):
+        with storage_tiering_configured():
+            IrodsController().restart(test_mode=True)
+            try:
+                with session.make_session_for_existing_admin() as admin_session:
+                    # Ensure that the delay queue is empty before creating the data object. This is to prevent false
+                    # negatives in a later assertion of the emptiness of the delay queue.
+                    admin_session.assert_icommand(["iqstat", "-a"], "STDOUT", "No delayed rules pending")
+
+                    self.user1.assert_icommand(create_command, "STDOUT")
+
+                    # Immediately check to see whether a restage was scheduled (it should not have been). We do this
+                    # first to ensure that the delay rule was not scheduled and processed by the time we check.
+                    admin_session.assert_icommand(["iqstat", "-a"], "STDOUT", "No delayed rules pending")
+
+                    # Then, make sure that the object is in the correct tier and has an access_time.
+                    self.assertTrue(lib.replica_exists_on_resource(self.user1, logical_path, self.tier2))
```

Comment:
Consider making the tier an argument that's passed in. It took me a few seconds to figure out that the test assumes the `create_command` targets tier 2.

Perhaps just a comment on the this helper function explaining that tier 2 must be the target resource is enough.

Reply:
Yep. It's more obvious now.